File: stripe/src/test/java/com/stripe/android/model/StripeModelTest.java
package com.stripe.android.model;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNotSame;
import static org.junit.Assert.assertTrue;

public class StripeModelTest {

    @Test
    public void equals_whenEquals_returnsTrue() {
        assertTrue(StripeModel.class.isAssignableFrom(Card.class));

        Card firstCard = Card.fromString(CardTest.JSON_CARD_USD);
        Card secondCard = Card.fromString(CardTest.JSON_CARD_USD);

        assertEquals(firstCard, secondCard);
        // Just confirming for sanity
        assertNotSame(firstCard, secondCard);
    }

    @Test
    public void equals_whenNotEquals_returnsFalse() {
        assertTrue(StripeModel.class.isAssignableFrom(Card.class));
        final Card firstCard = Card.create("4242", null, null, null);
        final Card secondCard = Card.create("4343", null, null, null);
        assertNotEquals(firstCard, secondCard);
    }

    @Test
    public void hashCode_whenEquals_returnsSameValue() {
        assertTrue(StripeModel.class.isAssignableFrom(Card.class));

        Card firstCard = Card.fromString(CardTest.JSON_CARD_USD);
        Card secondCard = Card.fromString(CardTest.JSON_CARD_USD);
        assertNotNull(firstCard);
        assertNotNull(secondCard);

        assertEquals(firstCard.hashCode(), secondCard.hashCode());
    }

    @Test
    public void hashCode_whenNotEquals_returnsDifferentValues() {
        assertTrue(StripeModel.class.isAssignableFrom(Card.class));

        Card usdCard = Card.fromString(CardTest.JSON_CARD_USD);
        Card eurCard = Card.fromString(CardTest.JSON_CARD_EUR);

        assertNotNull(usdCard);
        assertNotNull(eurCard);

        assertNotEquals(usdCard.hashCode(), eurCard.hashCode());
    }
}

File: stripe/src/test/java/com/stripe/android/model/StripeModelTest.kt
package com.stripe.android.model

import com.google.common.truth.Truth.assertThat
import com.stripe.android.model.parsers.CardJsonParser
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotSame
import kotlin.test.assertTrue
import org.json.JSONObject

/**
 * Test class for [StripeModel].
 */
class StripeModelTest {

    @BeforeTest
    fun setup() {
        assertTrue(StripeModel::class.java.isAssignableFrom(Card::class.java))
    }

    @Test
    fun equals_whenEquals_returnsTrue() {
        val firstCard = parse(CardTest.JSON_CARD_USD)
        val secondCard = parse(CardTest.JSON_CARD_USD)

        assertEquals(firstCard, secondCard)
        // Just confirming for sanity
        assertNotSame(firstCard, secondCard)
    }

    @Test
    fun equals_whenNotEquals_returnsFalse() {
        assertThat(CardParams("4242", 1, 2021))
            .isNotEqualTo(CardParams("4343", 1, 20201))
    }

    @Test
    fun hashCode_whenEquals_returnsSameValue() {
        assertThat(parse(CardTest.JSON_CARD_USD).hashCode())
            .isEqualTo(parse(CardTest.JSON_CARD_USD).hashCode())
    }

    @Test
    fun hashCode_whenNotEquals_returnsDifferentValues() {
        assertThat(CardFixtures.CARD_USD.hashCode())
            .isNotEqualTo(CardFixtures.CARD_EUR.hashCode())
    }

    private fun parse(json: JSONObject): Card {
        return requireNotNull(CardJsonParser().parse(json))
    }
}
